pup-0003.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Cherry-Picking Commits
+author: David Davis
+created: 01-Apr-2017
+status:
+---
+
+## Summary
+
+Currently in Pulp, we merge features to master. However, hotfixes and bug fixes
+are merged to release branches which then are merged to newer release branches
+and then to master (a process called "merging forward"). More info about
+merging forward can be found in our documentation:
+
+http://docs.pulpproject.org/en/nightly/dev-guide/contributing/merging.html#merging-your-changes
+
+Also, a good illustration of this process is outlined in the following webpage:
+
+http://nvie.com/posts/a-successful-git-branching-model/
+
+This PUP is to use cherry-picking instead of merging forward. Using this model,
+all PRs would be made against master and then changes would be cherry-picked to
+release branches as needed.
+
+
+## Motivation
+
+There are a few motivating factors for this proposal. First, our git history is
+complex with multiple branches feeding into one another and numerous merge
+commits. Using cherry-picks would simplify our git history by keeping it more
+linear and eliminating merge commits.
+
+Secondly, opening a PR requires knowledge of our current process. This can
+create a barrier to entry for community members. With the cherry-pick model,
+all PRs are opened against master and the onus to update release branches is on
+the person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+Another minor caveat is that changes that we want to only be made to older
+release branches (but not to newer ones) still need to be merged forward. We
+could elminate the need for this with cherry-picking. It's worth noting that
+this rarely happens.
+
+Lastly, we've also seen some mistakes where branches are accidentally merged
+into other branches. Commits could be accidentally cherry-picked but this would
+be less disastrous since it's only a single commit.
+
+
+## Detailed Design
+
+Here is an overview of the workflow:
+
+1. All PRs would be opened against master regardless of whether they contain a
+   feature, bug fix, or hotfix.
+2. After the PR is ready to be merged, the person merging the commit would
+   first merge the commit to master and then cherry-pick the commit to any
+   release branches that required the change. For a bugfix, this would be the
+   latest release branch. For hotfixes, this would be the oldest affected
+   release branch plus any newer branches.
```

Comment:
Good idea. I added a link to the branching doc of 2.12. I think that should provide a historical record of our branching scheme?